### PR TITLE
Introduce grace period before asking guests to verify their email address

### PR DESCRIPTION
**Pull request by barryhughes**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

In our latest release, we introduced a change that requires shoppers to verify their email address before seeing the order confirmation page, or order payment page, but only if WooCommerce is unable to recognize them.

Naturally, the desire was to make this as friction-free as possible and, for example, shoppers should not have to verify their email address immediately after checking out ... unfortunately, depending on the payment gateway in use, that became possible.

Therefore, this change introduces a grace period during which email verification will not be required. Further, the grace period is filterable so can be removed (or extended) as desired.

> :bulb: To clarify, this softens the requirement for email verification *for guest shoppers.* By itself, it does **not** touch the requirement for authenticated shoppers to be logged in before they can see their order confirmation.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. As a guest shopper, make a purchase.
2. After checking out, you should see the Order Received page.
3. Copy the URL and open this page in a new browser context (ie, an incognito mode tab or else a container tab, if you are using Firefox):
    - If we are within 10 minutes of the order being placed, then the Order Received page should be visible in both tabs (ie, the original tab where the shopper is 'known' and the separate tab where the shopper is unknown).
    - After 10 minutes, refreshing the second tab should result in you being asked to verify your email address (but in the original tab, the Order Received page should still be accessible).
4. There is no change for authenticated shoppers. In this case, if you repeat the above exercise, there is no grace period and if we cannot identify the shopper (as in the second browser context) they will be required to login.

### Notes

1. You don't have to wait 10 minutes when testing, as you can temporarily modify [this line of code](https://github.com/woocommerce/woocommerce/blob/35b85181e251a180486b5d884d5a80624bc13790/plugins/woocommerce/includes/shortcodes/class-wc-shortcode-checkout.php#L407) and replace `10 * MINUTE_IN_SECONDS` with `0` to simulate having waited 10 minutes. Or, if you don't like to touch core code, you could setup the following snippet in a suitable place (such as `wp-content/mu-plugins/test-39191.php`) then remove after testing:

```php
<?php
add_filter( 'woocommerce_order_email_verification_grace_period', '__return_zero' );
```

2. I am aware of this [related discussion](https://github.com/woocommerce/woocommerce/pull/38983). We still need to decide on the best course of action there (possibly following up with a second PR).

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [x] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement

#### Message 

Adds a grace period during which email verification will not be needed before the order confirmation (or payment) page is rendered.

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce/.wp-env.json
Comment (barryhughes) [inline plugins/woocommerce/.wp-env.json]: Since WP Env is a development tool, I think this is fine (and the idea is to make it available to the E2E tests). However, shout out if it's concerning.

**Comment by barryhughes** [comment]
- Requesting review, but let's not merge until we have agreement from all parties (for instance, we may want to request a further review from a payments-focused team).
- Also interested in non-technical perspectives on this (from a high-level, are there any concerns with this strategy?).
- Requested a Solaris review, partly in relation to my last point but also because of the E2E changes (maybe there is an existing and more elegant approach that I missed, regarding temporarily setting filters from E2E tests...or maybe that is simply a malpractice).

File: plugins/woocommerce/includes/shortcodes/class-wc-shortcode-checkout.php
Comment (vedanshujain) [inline plugins/woocommerce/includes/shortcodes/class-wc-shortcode-checkout.php:407]: should this be changed to hold stock minutes?

**Review by vedanshujain — APPROVED** [review]
Left a minor optional comment, I tested dby eleting the cookie to simulate the condition where the session won't have the email and worked as expected. Went again after 10 minutes, and as expected, I was required to verify the email.

Screenshot 2023-07-13 at 10 21 18 AM

File: plugins/woocommerce/includes/shortcodes/class-wc-shortcode-checkout.php
Comment (vedanshujain) [reply plugins/woocommerce/includes/shortcodes/class-wc-shortcode-checkout.php:407]: On second thought, might not be a good idea to change this, since these are two fundamentally different things.

Comment (barryhughes) [reply plugins/woocommerce/includes/shortcodes/class-wc-shortcode-checkout.php:407]: Yeah, they feel different to me, one being concerned with inventory management and the other access to historic information.

File: plugins/woocommerce/tests/e2e-pw/tests/shopper/checkout.spec.js
Comment (rodelgc) [inline plugins/woocommerce/tests/e2e-pw/tests/shopper/checkout.spec.js]: Since `clearFilters()` is an async function, we probably need to add an `await` to it? I'm just worried that the succeeding steps/tests might become flaky as we've had such instances before due to missing `await`'s.

File: plugins/woocommerce/tests/e2e-pw/tests/shopper/checkout.spec.js
Comment (rodelgc) [inline plugins/woocommerce/tests/e2e-pw/tests/shopper/checkout.spec.js]: Requesting for a minor change here: If you could please update the locator to `page.getByRole('heading', { name: 'Order received' })`, since we're transitioning the locators in our E2E tests to use user-facing attributes as a best practice mentioned [here](https://playwright.dev/docs/locators#locating-elements).

So, this line would look something like:
```js
await expect(
	page.getByRole( 'heading', { name: 'Order received' } )
).toBeVisible();

```

File: plugins/woocommerce/tests/e2e-pw/utils/filters.js
Comment (rodelgc) [inline plugins/woocommerce/tests/e2e-pw/utils/filters.js]: Same [review comment](https://github.com/woocommerce/woocommerce/pull/39191#discussion_r1264842597) I mentioned in the `setFilterValue` function about obtaining the `domain` value.

File: plugins/woocommerce/tests/e2e-pw/utils/filters.js
Comment (rodelgc) [inline plugins/woocommerce/tests/e2e-pw/utils/filters.js]: We run e2e tests against permanent test sites too, not just on localhost. So we couldn't hard-code `domain` to `localhost`. Instead, we should base its value on the base URL of the site the e2e tests are running against.

To achieve that, I would recommend this 2-step approach:
1. First, get the base URL of the test site. When we run the E2E tests against non-localhost sites, we set the base URL via the `BASE_URL` environment variable. If `BASE_URL` isn't specified, it defaults to `http://localhost:8086`. So we couldn't rely on the value of `BASE_URL` alone. A more accurate way to obtain the test site's base URL is by extracting it from the Playwright config file located in `plugins/woocommerce/tests/e2e-pw/playwright.config.js`. When you open that config file, you'd see a `use` object, with the `baseURL` property in it. That `baseURL` property takes care of getting the url accurately. You can then extract its value using a simple `require()`.
   GQtAcjNgWJ

3. Then, use the `new URL()` function to convert the extracted base URL string into a URL object, and then use the `hostname` property to obtain the value for the `domain`. 

Of course, feel free to use a different approach, like obtaining the domain via Javascript. As long as the end goal is to use the `domain` of the test site, whether it's localhost or not.

**Review by rodelgc — DISMISSED** [review]
Done with my review and left some comments. Thanks so much for making the necessary updates on the E2E tests @barryhughes! They tested well on `localhost`. I requested some changes so that E2E changes you made would also work on permanent test sites.

Summary of tests I did for this PR:
- ✔️ Ran the `allows guest customer to place an order` test just by itself
- ✔️ Ran `allows guest customer to place an order` twice in a row to ensure it's repeatable. I used the arguments ` --retries=0 --repeat-each=2 --workers=1`
- ✔️ Ran all tests in `shopper/checkout.spec.js`
- ⚠️  `allows guest customer to place an order` test failed when I ran all tests in `shopper/checkout.spec.js` against an external WooCommerce test site.

File: plugins/woocommerce/tests/e2e-pw/tests/shopper/checkout.spec.js
Comment (coreymckrill) [reply plugins/woocommerce/tests/e2e-pw/tests/shopper/checkout.spec.js]: Await added in 48fb4388ba

File: plugins/woocommerce/tests/e2e-pw/tests/shopper/checkout.spec.js
Comment (coreymckrill) [reply plugins/woocommerce/tests/e2e-pw/tests/shopper/checkout.spec.js]: Changed this line, and several others that were locating headings, to use `getByRole` in 48fb4388ba

File: plugins/woocommerce/tests/e2e-pw/utils/filters.js
Comment (coreymckrill) [reply plugins/woocommerce/tests/e2e-pw/utils/filters.js]: Used the recommended approach in 234323602f

File: plugins/woocommerce/tests/e2e-pw/utils/filters.js
Comment (coreymckrill) [reply plugins/woocommerce/tests/e2e-pw/utils/filters.js]: Used the recommended approach in 234323602f

**Comment by coreymckrill** [comment]
@rodelgc I'm taking over this PR since Barry is AFK. I've addressed your feedback, let me know what you think.

**Comment by nigeljamesstevenson** [comment]
Thanks a lot for making these updates @coreymckrill !

⚠️ `allows guest customer to place an order test` still failed when I ran all tests in `shopper/checkout.spec.js` against a JN.

Interestingly, it was failing on the line 349 assertion
```
		await expect( page.locator( 'h1.entry-title' ) ).toContainText(
		 	'Order received'
		);
```

but looked good when I updated it to
```

		await expect(
			page.getByRole( 'heading', { name: 'Order received' } )
		).toBeVisible();
```
(which you updated on a previous similar update in this PR) - so it certainly seem those locators are more stable..
I think it would be good to get this one updated also..

@rodelgc - any idea if this was the location at which your test was failing originally?

**Comment by coreymckrill** [comment]
🤔 @nigeljamesstevenson Line 349 is in a separate test, `allows existing customer to place order`, which isn't modified in this PR. Can you confirm which test is failing for you?

**Comment by nigeljamesstevenson** [comment]
> Line 349 is in a separate test, allows existing customer to place order, which isn't modified in this PR. Can you confirm which test is failing for you?

Doh! 🤦‍♂️ My bad @coreymckrill !  - It was indeed the  `allows existing customer to place order` that was failing for me. I will create a task for myself and take a peek at this. `allows guest customer to place an order` looks great!

**Comment by coreymckrill** [comment]
Since Nigel re-ran the tests and they are working, and we already have one approval here, I'm going to go ahead and merge.

**Comment by coreymckrill** [comment]
Actually, looks like I can't until @rodelgc confirms that requested changes were made.

**Review by lanej0 — APPROVED** [review]
Going to try another approval to get this out of "approval gridlock"

*Narrator:* the attempt didn't work.

**Comment by lanej0** [comment]
I did re-run the tests locally and against a Jurassic Ninja site and everything is good.  So I've unblocked this.

**Comment by coreymckrill** [comment]
Thank you!

**Pull request merged**
